cert-ldap-login-module-common/src/main/java/org/esbtools/auth/ldap/LdapRolesProvider.java

```diff
@@ -103,49 +127,78 @@ private void initializeFromConfiguration() throws Exception {
         // A flag that indicates whether to use the SO_KEEPALIVE socket option to attempt to more quickly detect when idle TCP connections have been lost or to prevent them from being unexpectedly closed by intermediate network hardware. By default, the SO_KEEPALIVE socket option will be used.
         options.setUseKeepAlive(ldapConfiguration.isKeepAlive());
 
-
-        if(ldapConfiguration.getUseSSL()) {
+        if (ldapConfiguration.getUseSSL()) {
             TrustStoreTrustManager trustStoreTrustManager = new TrustStoreTrustManager(
-                    ldapConfiguration.getTrustStore(),
-                    ldapConfiguration.getTrustStorePassword().toCharArray(),
-                    "JKS",
-                    true);
+                ldapConfiguration.getTrustStore(),
+                ldapConfiguration.getTrustStorePassword().toCharArray(),
+                "JKS",
+                true);
             SSLSocketFactory socketFactory = new SSLUtil(trustStoreTrustManager).createSSLSocketFactory();
 
-            ldapConnection = new LDAPConnection(
-                    socketFactory,
-                    options,
-                    ldapConfiguration.getServer(),
-                    ldapConfiguration.getPort(),
-                    ldapConfiguration.getBindDn(),
-                    ldapConfiguration.getBindDNPwd()
-            );
+            ldapConnection = new LDAPConnection(socketFactory, options);
         } else {
             LOGGER.warn("Not using SSL to connect to ldap. This is very insecure - do not use in prod environments!");
 
-            ldapConnection = new LDAPConnection(
-                    options,
-                    ldapConfiguration.getServer(),
-                    ldapConfiguration.getPort(),
-                    ldapConfiguration.getBindDn(),
-                    ldapConfiguration.getBindDNPwd()
-            );
+            ldapConnection = new LDAPConnection(options);
+        }
+
+        return ldapConnection;
+    }
+
+    private void connectIfNeeded() throws LDAPException {
+        if (connectionPool != null) {
+            return;
         }
 
-        BindRequest bindRequest = new SimpleBindRequest(ldapConfiguration.getBindDn(), ldapConfiguration.getBindDNPwd());
-        BindResult bindResult = ldapConnection.bind(bindRequest);
+        if (!readyForConnectionAttempt()) {
+            // It's too early to retry connecting again.
```

Comment:
I know we want to have automated connection retries after a certain time period (if no requests come in that trigger connectIfNeeded()), but if we're getting requests, why would we want to wait, as it is possible that the connection works the next time you try it before the specified time interval passes.

Reply:
If all requests wait, and it takes a while, we could fill thread pools server side and/or client side as lots of requests could pile up. It could also cause cascading failure if timeouts are hit by clients. It's the same idea as hystrix.

Reply:
Yeah I guess we wouldn't define a long interval anyway, so what you're saying makes sense.